src/ucp/core/ucp_am.c

```diff
@@ -533,161 +547,197 @@ ucp_am_handler(void *am_arg, void *am_data, size_t am_length,
 {
     ucp_worker_h worker = (ucp_worker_h)am_arg;
     ucp_am_hdr_t *hdr   = (ucp_am_hdr_t *)am_data;
-    uint16_t am_id      = hdr->am_hdr.am_id;
+    uint16_t am_id      = hdr->am_id;
 
     return ucp_am_handler_common(worker, hdr + 1, sizeof(*hdr), am_length,
                                  NULL, am_id, am_flags);
 }
 
-static ucp_am_unfinished_t *
-ucp_am_find_unfinished(ucp_worker_h worker, ucp_ep_h ep,
-                       ucp_ep_ext_proto_t *ep_ext,
-                       ucp_am_long_hdr_t *hdr, size_t am_length)
+static UCS_F_ALWAYS_INLINE ucp_am_unfinished_t*
+ucp_am_find_unfinished(ucp_worker_h worker, ucp_ep_ext_proto_t *ep_ext,
+                       uint64_t msg_id)
 {
     ucp_am_unfinished_t *unfinished;
-    /* TODO make this hash table for faster lookup */
+
     ucs_list_for_each(unfinished, &ep_ext->am.started_ams, list) {
-        if (unfinished->msg_id == hdr->msg_id) {
+        if (unfinished->msg_id == msg_id) {
             return unfinished;
         }
     }
 
     return NULL;
 }
 
-static ucs_status_t
-ucp_am_handle_unfinished(ucp_worker_h worker,
-                         ucp_am_unfinished_t *unfinished,
-                         ucp_am_long_hdr_t *long_hdr,
-                         size_t am_length, ucp_ep_h reply_ep)
+static UCS_F_ALWAYS_INLINE ucp_am_unfinished_t*
+ucp_am_alloc_unfinished(ucp_worker_h worker, ucp_ep_ext_proto_t *ep_ext,
+                        uint64_t msg_id)
+{
+    ucp_am_unfinished_t *unfinished;
+
+    unfinished = ucs_malloc(sizeof(ucp_am_unfinished_t), "ucp multi-frag AM");
```

Comment:
are there plans to optimize this?

Reply:
yes, I was going to define an mpool in the next PR